From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Jan Alexander Steffens (heftig)" <heftig@archlinux.org>
Date: Thu, 28 Aug 2025 00:18:01 +0200
Subject: [PATCH] libsysprof-capture: Disallow unloading the capture library

When a shared library with a built-in libsysprof-capture gets unloaded,
the `sysprof_collector_free` function which was registered as a TSD
destructor with `pthread_key_create` will also vanish, causing crashes
when threads terminate.

See: https://gitlab.freedesktop.org/mesa/mesa/-/issues/13571
---
 src/libsysprof-capture/meson.build | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/libsysprof-capture/meson.build b/src/libsysprof-capture/meson.build
index 828b8ed2fef5..7a06ed257f92 100644
--- a/src/libsysprof-capture/meson.build
+++ b/src/libsysprof-capture/meson.build
@@ -48,33 +48,41 @@ libsysprof_capture_deps = [
   dependency('threads'),
 ]
 
+libsysprof_capture_link_args = []
+if cc.has_link_argument('-Wl,-z,nodelete')
+  libsysprof_capture_link_args += ['-Wl,-z,nodelete']
+endif
+
 libsysprof_capture = static_library(
   'sysprof-capture-@0@'.format(libsysprof_capture_api_version),
   (libsysprof_capture_sources +
    mapped_ring_buffer_sources),
 
            dependencies: libsysprof_capture_deps,
                  c_args: [ '-DSYSPROF_CAPTURE_COMPILATION' ],
+              link_args: libsysprof_capture_link_args,
                 install: install_static,
   gnu_symbol_visibility: 'hidden',
                     pic: true,
 )
 
 libsysprof_capture_include_dirs = include_directories('.')
 
 libsysprof_capture_dep = declare_dependency(
            link_whole: libsysprof_capture,
          dependencies: libsysprof_capture_deps,
   include_directories: libsysprof_capture_include_dirs,
+            link_args: libsysprof_capture_link_args,
 )
 meson.override_dependency('sysprof-capture-@0@'.format(libsysprof_capture_api_version), libsysprof_capture_dep)
 
 if install_static
   pkgconfig.generate(
                  libsysprof_capture,
         subdirs: [ sysprof_header_subdir ],
     description: 'The static capture library for tools that generate profiling capture data',
       variables: [ 'datadir=' + datadir_for_pc_file ],
+    libraries: libsysprof_capture_link_args,
     libraries_private: libsysprof_capture_deps,
   )
 endif
